bebraven/platform
# New Sync to LMS flow using FormAssembly entered Salesforce data - Core Logic

Implements core functionality of the new Sync To LMS flow. There are TODOs littered throughout with corresponding Asana tasks for a future iteration.

# High level context

## The old Sync To LMS flow was:
1. Fellow or LC creates an account on the Join server
2. fills out an application
3. we store that information locally and push most of it to Salesforce
4. regional staff review, accept, confirm and map them to a cohort on Salesforce
5. regional staff hit a Sync To LMS button on Salesforce which calls into the Join server with the Campaign ID to sync
6. Join server pulls info from the Salesforce Campaign and combines it with information stored locally to hit the Canvas API and provision the users.

## The new flow for getting fully confirmed and enrolled Fellows and LCs into the Portal:
1. Fellow or LC fills out a FormAssembly form in Salesforce which populates the new Salesforce "BluePrint" data model
2. Regional staff funnel their application through to the point of creating a Participant object in a Program which represents someone who is fully ready to have Portal access and participate in the program
3. A nightly cron job provisions Participants in the Portal with the correct access
4. An on-demand Sync To LMS button is also available on Salesforce which kicks off the sync cronjob for that Program (aka course)
5. A launch logistics email goes out with a link to create a Braven account and get Portal access (the email must match)
6.  When they create the account, as long as Sync To LMS has been run they get Portal access

# How to test / Try this out
1. See the `README` for more info about setting up Salesforce. But here is the test Program that I created in our staging environment 
![image](https://user-images.githubusercontent.com/5596986/79499753-72ede900-7ff9-11ea-8bea-18efba2127e1.png)
2. Clicking on the Fellows tab you can see that I added an Enrolled Participant 
![image](https://user-images.githubusercontent.com/5596986/79499882-9b75e300-7ff9-11ea-89fb-fe4d34409708.png)
3. Did the same for Leadership Coaches.
4. Assuming you have the Salesforce ENV vars set in your shell, you can go here to run the sync [http://platformweb:3020/salesforce/sync_to_lms] by putting in the course ID that is set as the `Target Course In LMS` field on your test Salesforce Program
![image](https://user-images.githubusercontent.com/5596986/79500273-2f47af00-7ffa-11ea-81f4-31b26a19bcef.png)
5. You can also run it using a rake task. E.g. `bundle exec rake salesforce:sync_to_lms[71]`
6. Also, if you don't want to deal with your Canvas dev environment, you can change the `CANVAS_URL` and `CANVAS_TOKEN` ENV vars to the staging values.
